3scaleAdapter/pkg/threescale/proxy_conf.go

```diff
@@ -0,0 +1,261 @@
+type ProxyConfigCache struct {
+	limit                int
+	ttl                  time.Duration
+	refreshBuffer        time.Duration
+	flushWorkerRunning   bool
+	stopFlushWorker      chan bool
+	refreshWorkerRunning bool
+	stopRefreshWorker    chan bool
+	mutex                sync.RWMutex
+	cache                map[string]proxyStore
+}
+
+// NewProxyConfigCache returns a ProxyConfigCache
+// The accepted parameters are cacheTTL - the time between when an entry is added to the
+// cache and when it should expire, and limit - the total number of entries that can be
+// stored in the cache at any given time.
+func NewProxyConfigCache(cacheTTL time.Duration, refreshBuffer time.Duration, limit int) *ProxyConfigCache {
+	pcc := &ProxyConfigCache{
+		limit:         limit,
+		ttl:           cacheTTL,
+		refreshBuffer: refreshBuffer,
+		cache:         make(map[string]proxyStore, limit),
+	}
+	return pcc
+}
+
+// FlushCache iterates over the items in the cache and purges any expired items
+func (pc *ProxyConfigCache) FlushCache() {
+	pc.delete(pc.markForDeletion()...)
+}
+
+// RefreshCache iterates over the items in the cache and updates their values
+// If a configuration cannot be refreshed then the existing values is considered invalid
+// and will be purged from the cache
+func (pc *ProxyConfigCache) RefreshCache() {
+	forRefresh := pc.markForRefresh()
+	for _, store := range forRefresh {
+		cacheKey := pc.getCacheKeyFromCfg(store.replayWith.cfg)
+		pce, err := getFromRemote(store.replayWith.cfg, store.replayWith.client)
+		if err != nil {
+			log.Infof("error fetching from remote while refreshing cache for service id %s", store.replayWith.cfg.ServiceId)
+			pc.delete(cacheKey)
+			break
+		}
+		pc.set(cacheKey, pce, store.replayWith)
+	}
+}
+
+// StartFlushWorker starts a background process that periodically carries out the
+// functionality provided by FlushCache
+func (pc *ProxyConfigCache) StartFlushWorker() error {
+	if pc.flushWorkerRunning {
+		return errors.New("worker has already been started")
+	}
+
+	pc.stopFlushWorker = make(chan bool)
+	go pc.flushCache(pc.stopFlushWorker)
+	pc.flushWorkerRunning = true
+	return nil
+}
+
+// StopFlushWorker stops a background process started by StartFlushWorker if it has been started
+// Returns an error if the background task is not running
+func (pc *ProxyConfigCache) StopFlushWorker() error {
+	if pc.stopFlushWorker == nil || !pc.flushWorkerRunning {
+		return errors.New("worker has not been started")
+	}
+
+	pc.stopFlushWorker <- true
+	close(pc.stopFlushWorker)
+	pc.flushWorkerRunning = false
+	return nil
+}
+
+// StartRefreshWorker starts a background process that periodically carries out the
+// functionality provided by RefreshCache
+func (pc *ProxyConfigCache) StartRefreshWorker() error {
+	if pc.refreshWorkerRunning {
+		return errors.New("worker has already been started")
+	}
+
+	pc.stopRefreshWorker = make(chan bool)
+	go pc.refreshCache(pc.stopRefreshWorker)
+	pc.refreshWorkerRunning = true
```

Comment:
so if we have two threads that both want to `StartRefreshWorker` at the same time, is it possible that we end up launching parallel goroutines for refresh if the first thread is stopped right before writing to this variable?

Reply:
we would need for this to work an atomic boolean that we can test and set in the same operation (probably Go's libraries offer something already)